### ISSUE TITLE
dbWriteTable() throwing an "Invalid precision value" error with Access

### ISSUE THREAD

**Issue by DavisVaughan** [r-dbi/odbc#79]
### Issue Description and Expected Result

`dbWriteTable()` throwing an "Invalid precision value" error when used with Access.

Below I've given an example showing what I've tried/learned:
1) Both characters and numerics throw this error.
2) The error is stemming from `result_insert_dataframe`
3) The data types `DOUBLE` and `VARCHAR` seem to be correct, at least to create the table. Creating a new `odbcDataType.ACCESS` throws a different error that looks like it couldn't create the table with the other `SQL_DOUBLE` and `SQL_VARCHAR` data types I supplied.

### Database

Microsoft Access Version 14.0. 

<details>
<summary>Connection from odbc</summary>

``` r
library(DBI)
cn <- dbConnect(odbc::odbc(), dsn = "access-odbc")
dbGetInfo(cn)
#> $dbname
#> [1] "R:\\Life_Modeling_Team\\Vaughan\\1q16_NewBusiness_since_1q15.accdb"
#> 
#> $dbms.name
#> [1] "ACCESS"
#> 
#> $db.version
#> [1] "14.00.0000"
#> 
#> $username
#> [1] "admin"
#> 
#> $host
#> [1] ""
#> 
#> $port
#> [1] ""
#> 
#> $sourcename
#> [1] "access-odbc"
#> 
#> $servername
#> [1] "ACCESS"
#> 
#> $drivername
#> [1] "ACEODBC.DLL"
#> 
#> $odbc.version
#> [1] "03.80.0000"
#> 
#> $driver.version
#> [1] "Microsoft Access database engine"
#> 
#> $odbcdriver.version
#> [1] "03.51"
#> 
#> $supports.transactions
#> [1] FALSE
#> 
#> attr(,"class")
#> [1] "ACCESS"      "driver_info" "list"
```

</details>

### Other details

I am on a Windows computer running Windows 7 Professional.

### Reproducible Example

<details>

``` r
# Connecting (works, and gives warning)
library(DBI)
cn <- dbConnect(odbc::odbc(), dsn = "access-odbc")

# A simple data frame (One has a numeric, one has a character)
df_numeric <- data.frame(col_num  = 3)
df_char    <- data.frame(col_char = "hi")

# Write it. Uh oh. Both fail with the same error

dbWriteTable(cn, "table_num", df_numeric)
#> Error in result_insert_dataframe(rs@ptr, values): nanodbc/nanodbc.cpp:1791: HY104: [Microsoft][ODBC Microsoft Access Driver]Invalid precision value

dbWriteTable(cn, "table_char", df_char)
#> Error in result_insert_dataframe(rs@ptr, values): nanodbc/nanodbc.cpp:1944: HY104: [Microsoft][ODBC Microsoft Access Driver]Invalid precision value

# The tables were actually written, but it failed on the Insert of values
dbListTables(cn)
#>  [1] "MSysAccessStorage"          "MSysACEs"                  
#>  [3] "MSysComplexColumns"         "MSysIMEXColumns"           
#>  [5] "MSysIMEXSpecs"              "MSysNameMap"               
#>  [7] "MSysNavPaneGroupCategories" "MSysNavPaneGroups"         
#>  [9] "MSysNavPaneGroupToObjects"  "MSysNavPaneObjectIDs"      
#> [11] "MSysObjects"                "MSysQueries"               
#> [13] "MSysRelationships"          "MSysResources"             
#> [15] "1q16_ChubbNonPar"           "Sheet1$_ImportErrors"      
#> [17] "table_char"                 "table_num"

# The data type of the columns
dbDataType(cn, df_numeric)
#>  col_num 
#> "DOUBLE"
dbDataType(cn, df_char)
#>       col_char 
#> "VARCHAR(255)"

# I then created a new odbcDataType.ACCESS function
# Inspired by: 
# https://docs.microsoft.com/en-us/sql/odbc/microsoft/microsoft-access-data-types
`odbcDataType.ACCESS` <- function(con, obj, ...) {
  switch_type(obj,
    factor = "SQL_VARCHAR(255)",
    datetime = "SQL_TIMESTAMP",
    date = "SQL_TIMESTAMP",
    binary = "SQL_BINARY",
    integer = "SQL_INTEGER",
    double = "SQL_DOUBLE",
    character = "SQL_VARCHAR(255)",
    logical = "SQL_BINARY",
    list = "SQL_VARCHAR(255)",
    time = "SQL_TIMESTAMP",
    stop("Unsupported type", call. = FALSE)
  )
}

switch_type <- function(obj, ...) {
  switch(object_type(obj), ...)
}

object_type <- function(obj) {
  if (is.factor(obj)) return("factor")
  if (is(obj, "POSIXct")) return("datetime")
  if (is(obj, "Date")) return("date")
  if (is(obj, "blob")) return("binary")
  if (is(obj, "difftime")) return("time")

  return(typeof(obj))
}

# Let's remove the tables and try again.
# Seems like that SQL_DOUBLE / SQL_VARCHAR data types aren't actually supported, and that DOUBLE / VARCHAR was correct?

dbGetQuery(cn, "DROP TABLE table_num")
#> data frame with 0 columns and 0 rows

dbGetQuery(cn, "DROP TABLE table_char")
#> data frame with 0 columns and 0 rows

dbWriteTable(cn, "table_num", df_numeric)
#> Error in new_result(connection@ptr, statement) : 
#>  nanodbc/nanodbc.cpp:1344: 42000: [Microsoft][ODBC Microsoft Access Driver] Syntax error in field definition. 

dbWriteTable(cn, "table_char", df_char)
#> Error in new_result(connection@ptr, statement) : 
#>  nanodbc/nanodbc.cpp:1344: 42000: [Microsoft][ODBC Microsoft Access Driver] Syntax error in field definition. 
```

</details>

### Session info

<details>
<summary>Session Info</summary>

``` r
library(DBI)
library(RODBCDBI)
library(odbc)
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.4.0 (2017-04-21)
#>  system   i386, mingw32               
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  English_United States.1252  
#>  tz       America/New_York            
#>  date     2017-06-23
#> Packages -----------------------------------------------------------------
#>  package   * version date       source                      
#>  backports   1.1.0   2017-05-22 CRAN (R 3.4.0)              
#>  base      * 3.4.0   2017-04-21 local                       
#>  bit         1.1-12  2014-04-09 CRAN (R 3.4.0)              
#>  bit64       0.9-7   2017-05-08 CRAN (R 3.4.0)              
#>  blob        1.1.0   2017-06-21 Github (hadley/blob@9dd54d9)
#>  compiler    3.4.0   2017-04-21 local                       
#>  datasets  * 3.4.0   2017-04-21 local                       
#>  DBI       * 0.7     2017-06-18 CRAN (R 3.4.0)              
#>  devtools    1.13.2  2017-06-02 CRAN (R 3.4.0)              
#>  digest      0.6.12  2017-01-27 CRAN (R 3.4.0)              
#>  evaluate    0.10    2016-10-11 CRAN (R 3.4.0)              
#>  graphics  * 3.4.0   2017-04-21 local                       
#>  grDevices * 3.4.0   2017-04-21 local                       
#>  hms         0.3     2016-11-22 CRAN (R 3.4.0)              
#>  htmltools   0.3.6   2017-04-28 CRAN (R 3.4.0)              
#>  knitr       1.16    2017-05-18 CRAN (R 3.4.0)              
#>  magrittr    1.5     2014-11-22 CRAN (R 3.4.0)              
#>  memoise     1.1.0   2017-04-21 CRAN (R 3.4.0)              
#>  methods   * 3.4.0   2017-04-21 local                       
#>  odbc      * 1.1.0   2017-06-23 CRAN (R 3.4.0)              
#>  Rcpp        0.12.11 2017-05-22 CRAN (R 3.4.0)              
#>  rlang       0.1.1   2017-05-18 CRAN (R 3.4.0)              
#>  rmarkdown   1.6     2017-06-15 CRAN (R 3.4.0)              
#>  RODBC       1.3-15  2017-04-13 CRAN (R 3.4.0)              
#>  RODBCDBI  * 0.1.1   2016-03-14 CRAN (R 3.4.0)              
#>  rprojroot   1.2     2017-01-16 CRAN (R 3.4.0)              
#>  stats     * 3.4.0   2017-04-21 local                       
#>  stringi     1.1.5   2017-04-07 CRAN (R 3.4.0)              
#>  stringr     1.2.0   2017-02-18 CRAN (R 3.4.0)              
#>  tibble      1.3.1   2017-05-17 CRAN (R 3.4.0)              
#>  tools       3.4.0   2017-04-21 local                       
#>  utils     * 3.4.0   2017-04-21 local                       
#>  withr       1.0.2   2016-06-20 CRAN (R 3.4.0)              
#>  yaml        2.1.14  2016-11-12 CRAN (R 3.4.0)
```

</details>

**Comment by jimhester** [comment]
I don't think the issue is with the `odbcDataType`, the table is created ok, the error occurs when binding parameters to the table. If the default `odbcDataType` method was invalid there would be an error when creating the table.

Unless you want to get familiar the gdb debugger and try to track it down yourself I will likely have to setup an access database to figure out what is going wrong with the parameter binding. However that is unlikely to be done in the short term.

And Thank you for the detailed and reproducible example!

**Comment by DavisVaughan** [comment]
I agree that the issue is not with `odbcDataType`, that is what I found too.

On Monday, I'll try and give the gdb debugger a go. Haven't had to use it before, but I'll give it a shot. I'll post anything I find here. _Any quick resources to get up to speed on how it interacts with R?_ I found this video which looks decent? https://vimeo.com/11937905

You're welcome on the reproducible example! As a package co-author, I get the "help me help you" mentality!

**Comment by FelipeCarrillo** [comment]
Hi;
Would you be able to share the the code you used to connect to your database? I have an access database that I called 'test.accdb' on my desktop but I am getting the following error:
Error in config::get("test") : Config file config.yml not found in current working directory or parent directories.

I have the latest version of Rstudio, DBI, and config packages installed

**Comment by r2evans** [comment]
@jimhester  @DavisVaughan 

Any progress on this? I'm not familiar enough with gdb and R to go that step, is there anything I can do to help? There seem to be several problems with ODBC access to ACCDBs, it would be nice to be able to smooth some of them out.

**Comment by vh-d** [comment]
I thought I would place `Rcpp::Rcout` somewhere into the `odbc_result::bind_list()` function to see the statements before being sent to the database. Is there any easy way to do that?

**Comment by dpprdan** [comment]
I also gave it a shot, but I am basically having the same problem as @DavisVaughan, with subtle differences, though.

I am using the Office 2016 [Northwind sample db](https://docs.microsoft.com/en-US/dotnet/framework/data/adonet/sql/linq/downloading-sample-databases) and the 64bit odbc driver from the [Download Microsoft Access Database Engine 2016 Redistributable](https://www.microsoft.com/en-us/download/details.aspx?id=54920) on Windows 10.

``` r
library("DBI")
library("odbc")
nw_path <- normalizePath("D:/odbc_access/Northwind_2016.accdb")
cs <- paste0("Driver={Microsoft Access Driver (*.mdb, *.accdb)};Dbq=", nw_path, ";")
cn <- dbConnect(odbc::odbc(), .connection_string = cs)
```

Connection details

``` r
dbGetInfo(cn)
#> $dbname
#> [1] "D:\\odbc_access\\Northwind_2016.accdb"
#> 
#> $dbms.name
#> [1] "ACCESS"
#> 
#> $db.version
#> [1] "14.00.0000"
#> 
#> $username
#> [1] "admin"
#> 
#> $host
#> [1] ""
#> 
#> $port
#> [1] ""
#> 
#> $sourcename
#> [1] ""
#> 
#> $servername
#> [1] "ACCESS"
#> 
#> $drivername
#> [1] "ACEODBC.DLL"
#> 
#> $odbc.version
#> [1] "03.80.0000"
#> 
#> $driver.version
#> [1] "Microsoft Access database engine"
#> 
#> $odbcdriver.version
#> [1] "03.51"
#> 
#> $supports.transactions
#> [1] FALSE
#> 
#> attr(,"class")
#> [1] "ACCESS"      "driver_info" "list"
```

When writing a table I get an “Invalid precision value” error

``` r
dbWriteTable(cn, "mtcars", mtcars, overwrite = TRUE)
#> Error in result_insert_dataframe(rs@ptr, values): nanodbc/nanodbc.cpp:1944: HY104: [Microsoft][ODBC Microsoft Access Driver]Invalid precision value
```

Table is created, but no data written

``` r
dbReadTable(cn, "mtcars")
#>  [1] row_names mpg       cyl       disp      hp        drat      wt       
#>  [8] qsec      vs        am        gear      carb     
#> <0 rows> (or 0-length row.names)

dbDisconnect(cn)
```

To get to this point, I had to create an `odbcDataType.ACCESS()` function, though. Before I got the same error message as @patperu in his first example with #205. I have not figured out, why this is different from what @DavisVaughan reports, possibly 

```r
Error in new_result(connection@ptr, statement) : 
  nanodbc/nanodbc.cpp:1344: 42000: [Microsoft][ODBC Microsoft Access Driver] Syntax error in CREATE TABLE statement. 
```

I also used this information provided by Microsoft on [Access Data Types](https://docs.microsoft.com/en-us/sql/odbc/microsoft/microsoft-access-data-types?view=sql-server-2017), however contrary to @DavisVaughan, I used the `Data type (CREATETABLE)` instead of the `ODBC SQL data type`:

```r
#' @export
`odbcDataType.ACCESS` <- function(con, obj, ...) {
  switch_type(
    obj,
    factor = varchar(obj),
    datetime = "DATETIME",
    date = "DATE",
    time = "TIME",
    binary = "BINARY",
    integer = "INTEGER",
    double = "DOUBLE",
    character = varchar(obj),
    logical = "BIT",
    list = varchar(obj),
    stop("Unsupported type", call. = FALSE)
  )
}
```

Apparently Microsoft also [changes data types](https://support.office.com/en-us/article/data-types-for-access-desktop-databases-df2b83ba-cef6-436d-b679-3418f622e482) between Access versions, so supporting older Access DBs might be more involved?

I don't know how to work with the gdb debugger, so I am stuck here.

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> - Session info ----------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.5.2 (2018-12-20)
#>  os       Windows 10 x64              
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language EN                          
#>  collate  German_Germany.1252         
#>  ctype    German_Germany.1252         
#>  tz       Europe/Berlin               
#>  date     2019-02-24                  
#> 
#> - Packages --------------------------------------------------------------
#>  package     * version    date       lib source                     
#>  assertthat    0.2.0      2017-04-11 [1] CRAN (R 3.5.1)             
#>  backports     1.1.3      2018-12-14 [1] CRAN (R 3.5.1)             
#>  bit           1.1-14     2018-05-29 [1] CRAN (R 3.5.0)             
#>  bit64         0.9-7      2017-05-08 [1] CRAN (R 3.5.0)             
#>  blob          1.1.1      2018-03-25 [1] CRAN (R 3.5.1)             
#>  callr         3.1.1      2018-12-21 [1] CRAN (R 3.5.1)             
#>  cli           1.0.1.9000 2019-01-28 [1] Github (r-lib/cli@94e2fc5) 
#>  crayon        1.3.4      2017-09-16 [1] CRAN (R 3.5.1)             
#>  DBI         * 1.0.0      2018-05-02 [1] CRAN (R 3.5.1)             
#>  desc          1.2.0      2018-10-25 [1] Github (r-lib/desc@7c12d36)
#>  devtools      2.0.1      2018-10-26 [1] CRAN (R 3.5.1)             
#>  digest        0.6.18     2018-10-10 [1] CRAN (R 3.5.1)             
#>  evaluate      0.13       2019-02-12 [1] CRAN (R 3.5.2)             
#>  fs            1.2.6      2018-08-23 [1] CRAN (R 3.5.1)             
#>  glue          1.3.0      2018-07-17 [1] CRAN (R 3.5.1)             
#>  highr         0.7        2018-06-09 [1] CRAN (R 3.5.1)             
#>  hms           0.4.2      2018-03-10 [1] CRAN (R 3.5.1)             
#>  htmltools     0.3.6      2017-04-28 [1] CRAN (R 3.5.1)             
#>  knitr         1.21       2018-12-10 [1] CRAN (R 3.5.1)             
#>  magrittr      1.5        2014-11-22 [1] CRAN (R 3.5.1)             
#>  memoise       1.1.0      2017-04-21 [1] CRAN (R 3.5.1)             
#>  odbc        * 1.1.6      2019-02-12 [1] local                      
#>  pkgbuild      1.0.2      2018-10-16 [1] CRAN (R 3.5.1)             
#>  pkgconfig     2.0.2      2018-08-16 [1] CRAN (R 3.5.1)             
#>  pkgload       1.0.2      2018-10-29 [1] CRAN (R 3.5.1)             
#>  prettyunits   1.0.2      2015-07-13 [1] CRAN (R 3.5.1)             
#>  processx      3.2.1      2018-12-05 [1] CRAN (R 3.5.1)             
#>  ps            1.3.0      2018-12-21 [1] CRAN (R 3.5.2)             
#>  R6            2.4.0      2019-02-14 [1] CRAN (R 3.5.2)             
#>  Rcpp          1.0.0      2018-11-07 [1] CRAN (R 3.5.1)             
#>  remotes       2.0.2      2018-10-30 [1] CRAN (R 3.5.1)             
#>  rlang         0.3.1      2019-01-08 [1] CRAN (R 3.5.2)             
#>  rmarkdown     1.11       2018-12-08 [1] CRAN (R 3.5.1)             
#>  rprojroot     1.3-2      2018-01-03 [1] CRAN (R 3.5.1)             
#>  sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 3.5.1)             
#>  stringi       1.3.1      2019-02-13 [1] CRAN (R 3.5.2)             
#>  stringr       1.4.0      2019-02-10 [1] CRAN (R 3.5.2)             
#>  testthat      2.0.1      2018-10-13 [1] CRAN (R 3.5.1)             
#>  usethis       1.4.0      2018-08-14 [1] CRAN (R 3.5.1)             
#>  withr         2.1.2      2018-03-15 [1] CRAN (R 3.5.1)             
#>  xfun          0.5        2019-02-20 [1] CRAN (R 3.5.2)             
#>  yaml          2.2.0      2018-07-25 [1] CRAN (R 3.5.1)             
#> 
#> [1] D:/Users/Daniel/Documents/R/win-library/3.5
#> [2] C:/Program Files/R/R-3.5.2/library
```

</details>

**Comment by vh-d** [comment]
I think the issue is MS Access does not support SQLDescribeParam and https://github.com/r-dbi/odbc/commit/c2d2883ca47c8dd88457264071b4728c3f7881ba does not handle it well enough. I looked into source code of `pydbc` (which supports MS Access well) and they do parameter types more manually.

**Comment by vh-d** [comment]
I have reached a solution for Access. Users willing to test it can try

```
devtools::install_github("vh-d/odbc", ref = "batch_size_solution")
```
The downsize is that it data is inserted row-by-row in a loop and the results is even slower than `RODBC` package solution (which uses paste0() to bind data in an INSERT query)

**Comment by jimhester** [comment]
I wonder if this works better after https://github.com/r-dbi/odbc/pull/313/, which avoids using `SQLDescribeParam()` in some cases?

**Comment by divinenephron** [comment]
I can test it once #325 has been fixed.

**Comment by divinenephron** [comment]
The "Invalid precision value" error has been fixed as of e7619aa. Borrowing the good reprex from @dpprdan gives the following.

```r
library("DBI")
library("odbc")
nw_path <- normalizePath("Northwind_2016.accdb")
cs <- paste0("Driver={Microsoft Access Driver (*.mdb, *.accdb)};Dbq=", nw_path, ";")
cn <- dbConnect(odbc::odbc(), .connection_string = cs)
```

We can write a one-row `data.frame`:

```r
menu <- data.frame(item = c("Pizza"), price = c(12.00))
dbWriteTable(cn, "menu", menu, overwrite = TRUE)
dbReadTable(cn, "menu")
#>   item price
#> 1 Pizza    12
```

But we can't write a `data.frame` with two or more rows because we get a different error:

```r
dessert_menu <- data.frame(item = c("Ice cream", "Coffee"),
                           price = c(5.00, 4.50))
dbWriteTable(cn, "dessert_menu", dessert_menu, overwrite = TRUE)
#> Error in result_insert_dataframe(rs@ptr, values, batch_rows) : 
#>   nanodbc/nanodbc.cpp:1611: HYC00: [Microsoft][ODBC Microsoft Access Driver]Optional feature
#> not implemented
```

This is because the Microsoft Access Database Engine 2016 Redistributable doesn't support SQL statements which write multiple rows. This is a limitation of the ODBC driver being used, and not a limitation of the `odbc` library. Users can still loop through data frame one row at a time to write them:

```r
purrr::pwalk(dessert_menu,
             ~ dbWriteTable(cn, "dessert_menu", data.frame(...), append = TRUE))
dbReadTable(cn, "dessert_menu")
#>    item price
#> 1 Ice cream   5.0
#> 2    Coffee   4.5
```

Somebody could request a feature for the `odbc` library to elegantly work around this limitation.

<details>
<summary> devtools::session_info() </summary>

```
- Session info ------------------------------------------------------------
 setting  value                       
 version  R version 3.6.0 (2019-04-26)
 os       Windows 10 x64              
 system   x86_64, mingw32             
 ui       RStudio                     
 language (EN)                        
 collate  English_United Kingdom.1252 
 ctype    English_United Kingdom.1252 
 tz       Europe/London               
 date     2019-12-09                  

- Packages ----------------------------------------------------------------
 package     * version    date       lib source                     
 assertthat    0.2.1      2019-03-21 [1] CRAN (R 3.6.0)             
 backports     1.1.5      2019-10-02 [1] CRAN (R 3.6.1)             
 bit           1.1-14     2018-05-29 [1] CRAN (R 3.6.0)             
 bit64         0.9-7      2017-05-08 [1] CRAN (R 3.6.0)             
 blob          1.2.0      2019-07-09 [1] CRAN (R 3.6.1)             
 callr         3.3.2      2019-09-22 [1] CRAN (R 3.6.1)             
 cli           1.1.0      2019-03-19 [1] CRAN (R 3.6.0)             
 crayon        1.3.4      2017-09-16 [1] CRAN (R 3.6.0)             
 DBI         * 1.0.0      2018-05-02 [1] CRAN (R 3.6.1)             
 desc          1.2.0      2018-05-01 [1] CRAN (R 3.6.0)             
 devtools      2.2.1      2019-09-24 [1] CRAN (R 3.6.1)             
 digest        0.6.23     2019-11-23 [1] CRAN (R 3.6.1)             
 ellipsis      0.3.0      2019-09-20 [1] CRAN (R 3.6.1)             
 fs            1.3.1      2019-05-06 [1] CRAN (R 3.6.0)             
 glue          1.3.1      2019-03-12 [1] CRAN (R 3.6.0)             
 hms           0.5.2      2019-10-30 [1] CRAN (R 3.6.1)             
 magrittr      1.5        2014-11-22 [1] CRAN (R 3.6.0)             
 memoise       1.1.0      2017-04-21 [1] CRAN (R 3.6.0)             
 odbc        * 1.2.1.9000 2019-12-09 [1] Github (r-dbi/odbc@e7619aa)
 pillar        1.4.2      2019-06-29 [1] CRAN (R 3.6.1)             
 pkgbuild      1.0.6      2019-10-09 [1] CRAN (R 3.6.1)             
 pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 3.6.1)             
 pkgload       1.0.2      2018-10-29 [1] CRAN (R 3.6.0)             
 prettyunits   1.0.2      2015-07-13 [1] CRAN (R 3.6.0)             
 processx      3.4.1      2019-07-18 [1] CRAN (R 3.6.1)             
 ps            1.3.0      2018-12-21 [1] CRAN (R 3.6.0)             
 R6            2.4.1      2019-11-12 [1] CRAN (R 3.6.1)             
 Rcpp          1.0.3      2019-11-08 [1] CRAN (R 3.6.1)             
 remotes       2.1.0      2019-06-24 [1] CRAN (R 3.6.0)             
 rlang         0.4.2      2019-11-23 [1] CRAN (R 3.6.1)             
 rprojroot     1.3-2      2018-01-03 [1] CRAN (R 3.6.0)             
 rstudioapi    0.10       2019-03-19 [1] CRAN (R 3.6.1)             
 sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 3.6.0)             
 testthat      2.3.1      2019-12-01 [1] CRAN (R 3.6.1)             
 tibble        2.1.3      2019-06-06 [1] CRAN (R 3.6.0)             
 usethis       1.5.1      2019-07-04 [1] CRAN (R 3.6.1)             
 vctrs         0.2.0      2019-07-05 [1] CRAN (R 3.6.1)             
 withr         2.1.2      2018-03-15 [1] CRAN (R 3.6.0)             
 zeallot       0.1.0      2018-01-28 [1] CRAN (R 3.6.0)             

[1] C:/Users/XXXXXXX/Documents/R/win-library/3.6
[2] C:/Program Files/R/R-3.6.0/library
```
</details>

**Comment by vh-d** [comment]
We have solved this issue (#263) already. In my fork [here](https://github.com/vh-d/odbc/tree/batch_size_solution) and recently Jim came with his own implementation via ed2bf38d285c9ca3512c4c282cc6337addf2095d

Just set `options(odbc.batch_rows = 1)` or pass `batch_rows = 1` to `dbWriteTable()`.

BTW, it is not that Access does not accept INSERT statements with multiple rows. It does not support binding array of parameters. Moreover `RODBC`, which uses multi-line INSERTs is slightly faster than `odbc` in writing to Access on my machine.

**Comment by jimhester** [comment]
Ok, it sounds like this is fixed then, thanks for trying it out @divinenephron and @vh-d!
